curvefs/src/mds/space/volume_space.h

```diff
@@ -0,0 +1,186 @@
+        const std::vector<BlockGroup>& blockGroups) override;
+
+    /**
+     * @brief Remove all block groups and persistent records that belong to
+     *        current volume
+     * @return return SpaceOk if success, otherwise return error code
+     */
+    SpaceErrCode RemoveAllBlockGroups();
+
+    SpaceUsage* GetSpaceUsage() const { return usage_.get(); }
+
+ private:
+    VolumeSpace(uint32_t fsId,
+                uint64_t size,
+                uint32_t blockSize,
+                uint64_t blockGroupSize,
+                BitmapLocation location,
+                BlockGroupStorage* storage);
+
+ private:
+    SpaceErrCode AllocateBlockGroupsInternal(
+        uint32_t count,
+        const std::string& owner,
+        std::vector<BlockGroup>* blockGroups);
+
+    uint32_t AllocateFromAvailableGroups(uint32_t count,
+                                         const std::string& owner,
+                                         std::vector<BlockGroup>* groups);
+
+    uint32_t AllocateFromCleanGroups(uint32_t count,
+                                     const std::string& owner,
+                                     std::vector<BlockGroup>* groups);
+
+    SpaceErrCode AcquireBlockGroupInternal(uint64_t blockGroupOffset,
+                                           const std::string& owner,
+                                           BlockGroup* group);
+
+ private:
+    // persist block group to backend storage
+    SpaceErrCode PersistBlockGroup(const BlockGroup& group);
+    SpaceErrCode PersistBlockGroups(const std::vector<BlockGroup>& blockGroups);
+
+    // remove corresponding block group from backend storage
+    SpaceErrCode ClearBlockGroup(const BlockGroup& group);
+
+ private:
+    const uint32_t fsId_;
+    const uint32_t blockSize_;
+    const uint64_t blockGroupSize_;
+    const BitmapLocation bitmapLocation_;
+
+    uint64_t volumeSize_;
+
+    mutable bthread::Mutex mtx_;
+
+    // block groups are divided into three types
+    // 1. allocated
+    //    these block groups are now owned by clients(curve-fuse or
+    //    curvefs-metaserver), these block groups are forbidden to reallocate to
+    //    other clients.
+    //    allocated block groups are persisted into storage.
+    // 2. available
+    //    these block groups had been used by some clinets, but now they don't
+    //    have owners, and their space is partial used. so they can be
+    //    reallocated to other clients.
+    //    available block groups are persisted into storage, but doesn't have to
+    //    do this, we persist these block groups for two reasons, first, we can
+    //    statistics space usage more convenient, second is we can reallocate
+    //    these block groups to its previous owner.
+    // 3. clean
+    //    these block groups' space is never used, and they can be allocated to
+    //    other clients.
+
+    // key is block group offset
+    std::unordered_map<uint64_t, BlockGroup> allocatedGroups_;
```

Comment:
each block group contains 3 uint64 values, 2 uint32 values, and an optional string, it's much smaller.

for example, 1 TiB volume, 128MiB block group, memory usage is 
```
(1 << 40) / (128 << 20) * 64 / 1024 = 512K
```

and if all mountpoints it cleared, this will also be cleared from memory.